File: app/schemas/workflow.py
from dataclasses import dataclass
from typing import Optional

from .execution import ExecutionTrace
from .log import StructuredLogs
from .repair import RepairSuggestion
from .report import FinalReport
from .repository import DependencyGraph, RepoContext
from .root_cause import RootCauseResult
from .validation import ValidationResult


@dataclass
class WorkflowState:
    repo_url: Optional[str] = None
    ci_log: Optional[str] = None

    structured_logs: Optional[StructuredLogs] = None
    root_cause: Optional[RootCauseResult] = None
    repo_context: Optional[RepoContext] = None

    dependency_graph: Optional[DependencyGraph] = None
    fix_suggestion: Optional[RepairSuggestion] = None
    validation_result: Optional[ValidationResult] = None
    final_report: Optional[FinalReport] = None
    execution_trace: Optional[ExecutionTrace] = None

File: frontend/src/types/workflow.ts
export interface RootCauseResult {
  cause: "dependency_failure" | "build_error" | "test_failure" | "runtime_error" | "unknown";
  confidence: number;
  explanation: string;
}

export interface PatchChange {
  search: string;
  replace: string;
}

export interface PatchFile {
  path: string;
  changes: PatchChange[];
}

export interface RepairSuggestion {
  files: PatchFile[];
  reasoning: string;
  confidence: number;
}

export interface ValidationResult {
  valid: boolean;
  ci_status: string;
  risk: string;
  success_probability: number;
}

export interface ExecutionStep {
  name: string;
  status: string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  output?: any;
  error_message?: string | null;
  exception_type?: string | null;
  details?: string | null;
}

export interface ExecutionTrace {
  steps: ExecutionStep[];
}

export interface DependencyGraph {
  nodes: string[];
  edges: [string, string][];
}

export interface RepoContext {
  files: Record<string, string>;
}

export interface FinalReport {
  summary: string;
  root_cause: RootCauseResult;
  fix: RepairSuggestion;
  validation: ValidationResult;
  execution_trace?: ExecutionTrace | null;
  dependency_graph?: DependencyGraph | null;
  repo_context?: RepoContext | null;
}
